Enable a fast path for (static) qlinear for AArch64 through ACL directly.

fadara01 opened the pull request:
This enables a fast path for eager mode statically quantized matmuls for AArch64 through Arm Compute Library (ACL) directly.

PR #145942 addressed the high overhead in `qlinear_dynamic` on AArch64 (due to redundant weight pre-transpositions and reductions) by enabling a path that calls ACL directly.
This does the same thing and addresses the same overheads for (static) `qlinear`.

I benchmarked this PR (ACL direct integration for static quantization in ATen) against the current state of PyTorch (with #147498 which updates oneDNN to v3.7 included because it's a much stronger baseline than the current oneDNN version in PyTorch which is v3.5.3). See benchmarking script below.
My benchmark runs statically quantized linears for all combinations of `M = [8, 16, ..., 512]`, `K = [768, 1024, 2048, 4096]`, `N = [768, 1024, 2048, 4096]`. 

This PR gives an average speedup of **2x** for signed activations (`s8s8s8`) and **95x** for unsigned activations (`u8s8u8`) on a Neoverse-V1 with 16 threads.
The astronomical speedup for unsigned activation is because oneDNN v3.7 does not have an optimized implementation for `u8s8u8` on AArch64.

```
# SPDX-FileCopyrightText: Copyright 2025 Arm Limited and/or its affiliate <open-source-office@arm.com>
# SPDX-License-Identifier: BSD-3-Clause
import torch
import torch.nn as nn
from torch.quantization import QConfig
from torch.ao.quantization.observer import HistogramObserver, default_weight_observer
import torch
import torch.nn as nn
import numpy as np
import random
from argparse import ArgumentParser
import time

class ModelArgumentParser(ArgumentParser):
    def __init__(self) -> None:
        super().__init__()
        self.add_argument("--M",
                            help="M dimension",
                            type=int,
                            default=64
        )
        self.add_argument("--K",
                            help="K dimension",
                            type=int,
                            default=64
        )
        self.add_argument("--N",
                            help="N dimension",
                            type=int,
                            default=64
        )
        self.add_argument("--signed_input",
                            help="Use (signed) torch.qint8 for inputs instead of (unsigned) torch.quint8",
                            action="store_true"
        )
        self.add_argument("--seed",
                          help="Random seed",
                          type=int,
                          default=42
        )
        self.add_argument("--iters",
                          help="benchmark iterations",
                          default=500)

def set_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)

class LinearModel(nn.Module):
    def __init__(self, K, N):
        super(LinearModel, self).__init__()
        self.quant = torch.quantization.QuantStub()
        self.fc = nn.Linear(K, N)
        self.dequant = torch.quantization.DeQuantStub()

    def forward(self, x):
        x = self.quant(x)
        x = self.fc(x)
        x = self.dequant(x)
        return x

def quantize_model(model, args):
    qconfig = QConfig(
            activation=HistogramObserver.with_args(reduce_range=False,
            dtype=torch.qint8 if args.signed_input else torch.quint8),
            weight=default_weight_observer,
    )
    # Prepare the model for static quantization
    # Specify quantization configurations
    model.qconfig = qconfig
    model_prepared = torch.quantization.prepare(model_fp32)

    # Calibrate the model with sample inputs
    # Example input data for calibration
    with torch.no_grad():
        sample_data = torch.randn(args.M, args.K)
        model_prepared(sample_data)
    # Convert the prepared model to a quantized model
    model_quantized = torch.quantization.convert(model_prepared)
    return model_quantized


if __name__ == "__main__":
    parser = ModelArgumentParser()
    args = parser.parse_args()
    
    set_seed(args.seed)
    model_fp32 = LinearModel(args.K, args.N)
    model_quantized = quantize_model(model_fp32, args)
    
    inputs = torch.randn(args.M, args.K)
    times = []
    with torch.no_grad():
        # warmup
        for _ in range(10):
            model_quantized(inputs)
        # benchmark
        for _ in range(args.iters):
            s = time.time_ns()
            model_quantized(inputs)
            times.append((time.time_ns() - s) / 1e6)

    print("M,K,N,signed = ", args.M, args.K, args.N, args.signed_input)
    print("Min Times (ms) = ", min(times))
    print("Mean Times (ms) = ", np.mean(times))
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @malfet @snadampal @milpuz01

fadara01 committed:
Enable fast qlinear_dynamic path for AArch64 through ACL directly

This enables a fast path for eager mode dynamic quantization for AArch64 through Arm Compute Library (ACL) directly.

Context: PR #126687 enabled an optimized implementation for qlinear_dynamic for aarch64 through ideep → oneDNN → ACL which improved performance by ~10x compared to the previous implementation.
However, the current qlinear_dynamic path (ideep → oneDNN → ACL) suffers from high overhead due to the API friction between the stateless oneDNN API and the stateful ACL low-precision GEMM (lowp_gemm) API - for example, ACL's lowp_gemm objects cache information like weights reduction or weights in optimized memory format which oneDNN does not allow due to its stateless nature.
Hence, ACL currently runs a (redundant) sum of columns and pre-transposition (to the gemm kerne's optimal format) for each GEMM operation.
This PR addresses the sub-optimalities above by integrating ACL directly with qlinear_dynamic. This approach yields an average speedup (averaged over context_lengths of 2^3 up to 2^9) of ~ 50% for bert-base-uncased, bert-large-uncased, roberta-base, distilbert-base-uncased with 16 threads on a Neoverse-V1 (with transformers==4.48).
To achieve this we introduce PackedLinearWeightsACL (as a subclasses of PackedLinearWeightsOnednn ) with an implementation of qlinear_dynamic that uses ACL directly, while qlinear still follows the oneDNN path.

fadara01 committed:
Enable a fast path for (static) qlinear for AArch64 through ACL directly.

This enables a fast path for eager mode static quantization for AArch64 through Arm Compute Library (ACL) directly.

PR #145942 addressed the high overhead in qlinear_dynamic on AArch64 (due to redundant weight pretranspositions and reductions) by enabling a path that calls ACL directly.
This does the same thing but for (static) qlinear.

fadara01 commented:
@pytorchbot label "module: arm"

fadara01 commented:
@pytorchbot label "arm priority"

digantdesai commented on CMakeLists.txt:
```diff
@@ -1250,6 +1250,18 @@ if(USE_MIMALLOC)
+        include_directories(${ACL_INCLUDE_DIRS})
+        message(STATUS "ACL Include: ${ACL_INCLUDE_DIRS}")
+        message(STATUS "ACL Library: ${ACL_LIBRARIES}")
+    else()
```
Nit: too verbose?

fadara01 replied:
Removed the:
```
message(STATUS "ACL Include: ${ACL_INCLUDE_DIRS}")
message(STATUS "ACL Library: ${ACL_LIBRARIES}")
```

digantdesai commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -1107,6 +1108,97 @@ static at::Tensor linear_int8_with_onednn_weight(
+  }
+}
+
+at::Tensor PackedLinearWeightsACL::apply(
```
This will be yet another option in addition to existing alternatives to run this quantized operator on Arm CPUs. What are your thoughts, at a high level from maintenance point of view, on consolidating some of these options? Or even directly using Kleidi kernels here with `at::parallel` and other native utils?

fadara01 replied:
The motivation for this approach is to address the overheads arising from the incompatibilities between the oneDNN and ACL `lowp_gemm` APIs in the current path (without this PR): `PyTorch -> ideep -> oneDNN -> ACL` (See motivation on #145942)
oneDNN's API is stateless, while ACL's `lowp_gemm` API is stateful. The current path utilizes ACL's `lowp_gemm` operators from oneDNN. Given that oneDNN is stateless, ACL's `lowp_gemm` cannot have state and will need to run (redundant) weights pre-transpositions and reductions on every forward call. This PR enables ACL's `lowp_gemm` directly from PyTorch, which allows ACL to cache pre-transpositions and reductions for each layer, instead of recomputing them on every forward call.

> What are your thoughts, at a high level from maintenance point of view, on consolidating some of these options? Or even directly using Kleidi kernels here with at::parallel and other native utils?

Unfortunately, KleidiAI currently has minimal support for `int8` GEMMs in general, and no kernels for `int8` statically quantized GEMMs.
Once those kernels are in KleidiAI they will be part of ACL anyway and will be utilized by the path this PR enables. 
We will then benchmark the (direct) `PyTorch -> KleidiAI` path against the `PyTorch -> ACL -> KleidiAI` path. If it turns out that ACL is adding overhead, we'll integrate KleidiAI kernels directly here.

digantdesai replied:
I guess my concern was around maintainability, so many different ways to run this going through build and runtime selection just for Arm CPUs. While I understand the rationale, we should also aim to actively trim dead paths.

For example, I saw somewhere that it is possible to fallback to onednn when ACL path isn't valid, can we support those through ACL and remove onednn path.

fadara01 replied:
> we should also aim to actively trim dead paths.

Yes, I  agree with you. 
Currently, we're extending the existing `PackedLinearWeightsOneDNN` with our new `PackedLinearWeightsACL` rather than replacing it entirely. This is because ACL's support for per-channel quantized weights is still incomplete.  As you can see [here](https://github.com/pytorch/pytorch/pull/147337/files#diff-67b1929df06adddb95e57435d37276993efed63acb341101ff7decbd06c87d50R284), we only prepack to `PackedLinearWeightsACL` when weights are per-tensor affine quantized. Once ACL provides full support for per-channel quantized weights, we can drop the dependency on `PackedLinearWeightsOneDNN` and make ACL the default (and only) option for this path - i.e. `PackedLinearWeightsACL` extends `LinearPackedParamsBase` directly

> it is possible to fallback to onednn when ACL path isn't valid, can we support those through ACL and remove onednn path.

Regarding the fallback to `PackedLinearWeightsOneDNN::apply` or `PackedLinearWeightsOneDNN::apply_dynamic`, it's not strictly needed as I haven’t encountered any workload causing ACL's validation to fail. I included these fallbacks purely as a precaution. Since our implementation extends `PackedLinearWeightsOneDNN`, I wanted to guarantee that the behavior remains exactly the same as the previous implementation in all cases (plus the perf improvements)

digantdesai commented on aten/src/ATen/native/quantized/cpu/ACLUtils.h:
```diff
@@ -0,0 +1,524 @@
+        // PR: https://review.mlplatform.org/c/ml/ComputeLibrary/+/12820/8 This
+        // will be set to the actual src offset value at runtime.
+        arm_compute::QuantizationInfo(
+            1.0, std::numeric_limits<int8_t>::max(), true));
```
```suggestion
            /*scale=*/ 1.0, /*offset=*/ std::numeric_limits<int8_t>::max(), /*is_dynamic=*/ true));
```

fadara01 replied:
Done

digantdesai commented on aten/src/ATen/native/quantized/cpu/ACLUtils.h:
```diff
@@ -0,0 +1,524 @@
+
+namespace at::native::acl_utils {
+
+using ACLQuantMatmulCacheKey = std::tuple<
```
Nit: we are already in `acl_utils` namespace, do we still want to prefix everything with `ACL`?

fadara01 replied:
Good point!
I removed the ACL prefixes from all structs inside the `at::native::acl_utils` namespace.

digantdesai reviewed (commented):
Looks good at a high level. Left some comments.

Do we have existing python tests and CI setups which guarantee we test these new impls?

fadara01 commented:
> Do we have existing python tests and CI setups which guarantee we test these new impls?
Yes, current tests exercise this impl when `AT_MKLDNN_ACL_ENABLED` is set (it's set to `1` by default for AArch64 builds).
This PR fixes a currently (disabled on AArch64) but failing quantization test - #145216

malfet commented on CMakeLists.txt:
```diff
@@ -1250,6 +1250,16 @@ if(USE_MIMALLOC)
+if(USE_MKLDNN_ACL)
+    find_package(ACL REQUIRED)
+    if(ACL_FOUND)
+        include_directories(${ACL_INCLUDE_DIRS})
```
This is a really bad pattern, what's wrong with `target_include_directories(torch_cpu PRIVATE $(ACL_INCLUDE_DIRS})`?

fadara01 replied:
Ack, I removed this from the root `CMakeLists.txt` into the caffe2 one where `torch_cpu` is defined.
I removed the redundant `if(ACL_FOUND)` check and used `target_include_directories(torch_cpu PRIVATE $(ACL_INCLUDE_DIRS})` instrad of `include_directories(${ACL_INCLUDE_DIRS})` as you suggested.

Does it look better now?

fadara01 commented:
Thanks for your feedback @digantdesai , @malfet 
I addressed your comments, could you please have another look?

cc: @milpuz01

fadara01 committed:
Enable Direct Use of Arm Compute Library (ACL) in ATen

ACL is already built with PyTorch as a shared library when USE_MKLDNN_ACL is set.
Currently, it is only used indirectly in ATen via oneDNN for AArch64 targets. However there are cases where it makes sense to utilize ACL directly without  oneDNN as an intermediary - e.g. quantization. See #145942, #147337, #146620.
This patch enables such use cases by exposing ACL to ATen

fadara01 commented:
@malfet I created a standalone PR #148542 for all cmake related changes and removed them from here to ease the review process.

fadara01 commented:
Closing in favor of ghstack PR https://github.com/pytorch/pytorch/pull/148585 which has all comments addressed
